datafusion/optimizer/src/propagate_empty_relation.rs

```diff
@@ -0,0 +1,331 @@
+    use datafusion_common::{Column, ScalarValue};
+    use datafusion_expr::{
+        binary_expr, col, lit, logical_plan::builder::LogicalPlanBuilder, Expr, JoinType,
+        Operator,
+    };
+
+    use super::*;
+
+    fn assert_optimized_plan_eq(plan: &LogicalPlan, expected: &str) {
+        let rule = PropagateEmptyRelation::new();
+        let optimized_plan = rule
+            .optimize(plan, &mut OptimizerConfig::new())
+            .expect("failed to optimize plan");
+        let formatted_plan = format!("{:?}", optimized_plan);
+        assert_eq!(formatted_plan, expected);
+        assert_eq!(plan.schema(), optimized_plan.schema());
+    }
+
+    fn assert_together_optimized_plan_eq(plan: &LogicalPlan, expected: &str) {
+        let optimize_one = EliminateFilter::new()
+            .optimize(plan, &mut OptimizerConfig::new())
+            .expect("failed to optimize plan");
+        let optimize_two = PropagateEmptyRelation::new()
+            .optimize(&optimize_one, &mut OptimizerConfig::new())
+            .expect("failed to optimize plan");
+        let formatted_plan = format!("{:?}", optimize_two);
+        assert_eq!(formatted_plan, expected);
+        assert_eq!(plan.schema(), optimize_two.schema());
+    }
+
+    #[test]
+    fn propagate_empty() -> Result<()> {
+        let plan = LogicalPlanBuilder::empty(false)
+            .filter(Expr::Literal(ScalarValue::Boolean(Some(true))))?
+            .limit(10, None)?
+            .project(vec![binary_expr(lit(1), Operator::Plus, lit(1))])?
+            .build()?;
+
+        let expected = "EmptyRelation";
+        assert_optimized_plan_eq(&plan, expected);
+
+        Ok(())
+    }
+
+    #[test]
+    fn propagate_union_right_empty() -> Result<()> {
+        let table_scan = test_table_scan()?;
+        let left = LogicalPlanBuilder::from(table_scan).build()?;
+        let right_table_scan = test_table_scan_with_name("test2")?;
+        let right = LogicalPlanBuilder::from(right_table_scan)
+            .filter(Expr::Literal(ScalarValue::Boolean(Some(false))))?
+            .build()?;
+
+        let plan = LogicalPlanBuilder::from(left).union(right)?.build()?;
+
+        let expected = "Projection: a, b, c\
+            \n  TableScan: test";
+        assert_together_optimized_plan_eq(&plan, expected);
+
+        Ok(())
+    }
+
+    #[test]
+    fn propagate_union_left_empty() -> Result<()> {
+        let table_scan = test_table_scan()?;
+        let left = LogicalPlanBuilder::from(table_scan)
+            .filter(Expr::Literal(ScalarValue::Boolean(Some(false))))?
+            .build()?;
+        let right_table_scan = test_table_scan_with_name("test2")?;
+        let right = LogicalPlanBuilder::from(right_table_scan).build()?;
+
+        let plan = LogicalPlanBuilder::from(left).union(right)?.build()?;
+
+        let expected = "Projection: a, b, c\
+            \n  TableScan: test2";
+        assert_together_optimized_plan_eq(&plan, expected);
+
+        Ok(())
+    }
+
```

Comment:
Could you please add one more UT to cover the case that the union inputs have slightly different schema ?

Reply:
have added it.